core/jvm/src/unsafe/UnsafeBufferOperationsJvm.kt

```diff
@@ -0,0 +1,158 @@
+ * and the copy will be created if it could not be omitted.
+ *
+ * @param buffer a buffer to read from
+ * @param minimumCapacity the minimum amount of writable space
+ * @param writeAction an action that will be invoked on a [ByteBuffer] that will be added to a [buffer] by the end of
+ * the call
+ *
+ * @throws IllegalStateException when [minimumCapacity] is too large and could not be fulfilled.
+ *
+ * @sample kotlinx.io.samples.unsafe.UnsafeReadWriteSamplesJvm.readFromByteChannel
+ */
+@UnsafeIoApi
+public inline fun UnsafeBufferOperations.writeToTail(
+    buffer: Buffer,
+    minimumCapacity: Int,
+    writeAction: (ByteBuffer) -> Unit
+) {
+    writeToTail(buffer, minimumCapacity) { rawData, pos, limit ->
+        val bb = ByteBuffer.wrap(rawData, pos, limit - pos).slice()
+        writeAction(bb)
+        bb.position()
+    }
+}
+
+/**
+ * Provides read-only access to [buffer]'s data by filling provided [iovec] array with [ByteBuffer]'s representing
+ * [buffer]'s data, supplying it to [readAction] and consuming number of bytes returned by the [readAction].
+ *
+ * If there's not enough space in [iovec] to fit all byte buffers, only a prefix of [buffer]'s data will be supplied to
+ * [readAction].
+ * If the number of byte buffers available for read is less than [iovec]'s size,
+ * only a prefix of [iovec] will be filled.
+ *
+ * The second [readAction]'s parameter denotes the number of buffers supplied.
+ *
+ * The value returned by the [readAction] is interpreted as the number of consumed bytes.
+ * The size of the buffer will be reduced by that value,
+ * and the corresponding number of bytes from buffer's prefix will be no longer available for read.
+ * If data was not consumed, the [readAction] should return `0`.
+ *
+ * If [readAction] ends execution by throwing an exception, no data will be consumed from the buffer.
+ *
+ * If the [iovec] contains any references, it will be overridden during the call.
+ *
+ * If the [buffer] is empty, [IllegalArgumentException] will be thrown.
+ *
+ * The data is passed to the [readAction] directly from the buffer's internal storage without copying on
+ * the best effort basis, meaning that there are no strong zero-copy guarantees
+ * and the copy will be created if it could not be omitted.
+ *
+ * @param buffer a buffer to read from
+ * @param iovec a temporary array to store [ByteBuffer]s with data from [buffer]'s prefix
+ * @param readAction an action that will be invoked on an array filled with [ByteBuffer]s holding data from [buffer]'s
+ * prefix
+ *
+ * @throws IllegalArgumentException when the [buffer] is empty.
+ * @throws IllegalArgumentException when the [iovec] is empty.
+ *
+ * @sample kotlinx.io.samples.unsafe.UnsafeReadWriteSamplesJvm.gatheringWrite
+ *
+ */
+@UnsafeIoApi
+public inline fun UnsafeBufferOperations.readBulk(
+    buffer: Buffer,
+    iovec: Array<ByteBuffer?>,
+    readAction: (Array<ByteBuffer?>, Int) -> Long
```

Comment:
Same about named parameters and ULong

Reply:
See https://github.com/Kotlin/kotlinx-io/pull/334#discussion_r1652561311